### ISSUE TITLE
Access Denied with EnumerateFileSystemEntries

### ISSUE THREAD

**Issue by AlphaFS-Fan** [alphaleonis/AlphaFS#157]
I would like to write powershell script to change attributes of all files in given path. My script works fine, but if it finds a directory without access to it it stops with Access Denied exception. I tried try/catch exception handling but it didn't help. I also tried with DirectoryEnumerationOptions and   ContinueOnException without any luck.
This command makes Access Denied exception:

$AllFilesArray=[Alphaleonis.Win32.Filesystem.Directory]::EnumerateFileSystemEntries("C:\Path","_._",1)

Can you help how could I bypass this exception and make my script continue ?

Another matter. Is there a way to scan/enumerate the path without symbolic links ? I would like to change attriibutes for files except those in symlinked folders.

**Comment by Yomodo** [comment]
Regarding exception, does this work for you:

```
$Private:dirEnumOptions = [Alphaleonis.Win32.Filesystem.DirectoryEnumerationOptions]'Files, ContinueOnException'
$Private:pathFormat = [Alphaleonis.Win32.Filesystem.PathFormat]::FullPath

Foreach ($fse In [Alphaleonis.Win32.Filesystem.Directory]Directory::EnumerateFileSystemEntries('C:\Path', '*', $dirEnumOptions, $pathFormat)) {
    Write-Host $fse
}
```

To skip reparse points, try it with this:

```
$Private:dirEnumOptions = [Alphaleonis.Win32.Filesystem.DirectoryEnumerationOptions]'Files, ContinueOnException, SkipReparsePoints'
```

**Comment by AlphaFS-Fan** [comment]
Thanks, but doesn't work. First I took wrong method - EnumerateFileSystemEntries doesn't have DirectoryEnumerationOptions option.
Second - I figured I should use EnumerateFileSystemEntryInfos<T> method but don't know how. If I run 

Foreach ($fse In [Alphaleonis.Win32.Filesystem.Directory]::EnumerateFilesystemEntryInfos('C:\Path', '*', $dirEnumOptions, $pathFormat)) {

I get this error:
Cannot find an overload for "EnumerateFilesystemEntryInfos" and the argument count: "4".
At C:\admin\archive.ps1:6 char:18
- Foreach ($fse In [Alphaleonis.Win32.Filesystem.Directory]::EnumerateFilesystemEn ...
- ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  - CategoryInfo          : NotSpecified: (:) [], MethodException
  - FullyQualifiedErrorId : MethodCountCouldNotFindBest

**Comment by Yomodo** [comment]
Calling methods with generics requires a different approach.
This might help you:

```
Function Invoke-GenericMethod {
    Param(
        $Instance,
        [String]$MethodName,
        [Type[]]$TypeParameters,
        [Object[]]$MethodParameters
    )

    [Collections.ArrayList]$Private:parameterTypes = @{}
    ForEach ($Private:paramType In $MethodParameters) { [Void]$parameterTypes.Add($paramType.GetType()) }

    $Private:method = $Instance.GetMethod($methodName, "Instance,Static,Public", $Null, $parameterTypes, $Null)

    If ($Null -eq $method) { Throw ('Method: [{0}] not found.' -f ($Instance.ToString() + '.' + $methodName)) }
    Else {
        $method = $method.MakeGenericMethod($TypeParameters)
        $method.Invoke($Instance, $MethodParameters)
    }
}
```

And call it from PowerShell, like:

```
ForEach ($Private:folder In (Invoke-GenericMethod `
    -Instance           ([Alphaleonis.Win32.Filesystem.Directory]) `
    -MethodName         EnumerateFileSystemEntryInfos `
    -TypeParameters     Alphaleonis.Win32.Filesystem.FileSystemEntryInfo `
    -MethodParameters   'C:\', '*',
                        ([Alphaleonis.Win32.Filesystem.DirectoryEnumerationOptions]'Folders, SkipReparsePoints, ContinueOnException'),
                        ([Alphaleonis.Win32.Filesystem.PathFormat]::FullPath))) {

    Write-Host $folder.FullPath
}
```

**Comment by Alvalvar** [comment]
Thank you Yomodo! Function "Invoke-GenericMethod" worked well in folders where access is denied. But i don`t know in what catalog in directory access is denied... How i can get exception "access denied in $Path" and continue enumerate directory?

**Comment by Yomodo** [comment]
Hi! Unfortunately, _ContinueOnException_ currently does not keep track of encountered Exceptions. They are simply ignored.  Reporting Exceptions will hopefully be added in a future release.

**Comment by Yomodo** [comment]
Now available: 663e236b0d1ad7c31297565b68996050be5cabaa
